collector/mongod/collections_status.go

```diff
@@ -121,64 +122,58 @@ func (collStatList *CollectionStatList) Describe(ch chan<- *prometheus.Desc) {
 	collectionIndexesSize.Describe(ch)
 }
 
-var (
-	logSuppressCS = make(map[string]bool)
-)
+var logSuppressCS = make(map[string]struct{})
```

Comment:
Access to that global variable is not synchronized:

```
==================
WARNING: DATA RACE
Write at 0x00c00015f7a0 by goroutine 120:
  runtime.mapdelete_faststr()
      /usr/local/Cellar/go/1.13.8/libexec/src/runtime/map_faststr.go:297 +0x0
  github.com/percona/mongodb_exporter/collector/mongod.GetCollectionStatList()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongod/collections_status.go:141 +0x160
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).collectMongod()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:277 +0x36e
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).scrape()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:213 +0x69a
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).Collect()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:162 +0x46
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:442 +0x1f5

Previous write at 0x00c00015f7a0 by goroutine 121:
  runtime.mapdelete_faststr()
      /usr/local/Cellar/go/1.13.8/libexec/src/runtime/map_faststr.go:297 +0x0
  github.com/percona/mongodb_exporter/collector/mongod.GetCollectionStatList()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongod/collections_status.go:141 +0x160
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).collectMongod()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:277 +0x36e
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).scrape()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:213 +0x69a
  github.com/percona/mongodb_exporter/collector.(*MongodbCollector).Collect()
      /Users/aleksi/Code/Percona/PMM-2.0/src/github.com/percona/mongodb_exporter/collector/mongodb_collector.go:162 +0x46
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:442 +0x1f5

Goroutine 120 (running) created at:
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:534 +0x11d6
  github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/http.go:126 +0x7d
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:40 +0xe4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:100 +0xf8
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2387 +0x288
  net/http.serverHandler.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2802 +0xce
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:1890 +0x837

Goroutine 121 (running) created at:
  github.com/prometheus/client_golang/prometheus.(*Registry).Gather()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/registry.go:534 +0x11d6
  github.com/prometheus/client_golang/prometheus/promhttp.HandlerFor.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/http.go:126 +0x7d
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerInFlight.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:40 +0xe4
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1()
      /Users/aleksi/Soft/GOPATH/pkg/mod/github.com/prometheus/client_golang@v1.0.0/prometheus/promhttp/instrument_server.go:100 +0xf8
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2007 +0x51
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2387 +0x288
  net/http.serverHandler.ServeHTTP()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:2802 +0xce
  net/http.(*conn).serve()
      /usr/local/Cellar/go/1.13.8/libexec/src/net/http/server.go:1890 +0x837
==================
```

```
go install -race ./...
mongodb_exporter --log.level=debug --collect.database --collect.collection --collect.topmetrics --collect.indexusage --collect.connpoolstats
```

```
ab -n 1000 -c 10 http://localhost:9216/metrics
```

I'm not sure if it has to be global, or if we need that functionality at all, but race condition is definitely there.

Reply:
Confirmed and fixed